### DISCUSSION TITLE
Clarification about VB.NET

### DISCUSSION THREAD

**Discussion by jp2masa** [AvaloniaUI/Avalonia#14286] [General]
I noticed that #10387 (VB.NET support) was locked. I understand that the discussion was basically stuck, but I think the locking feature was a bit abused.

What actually concerns me is that the [2023 Stack Overflow Developer Survey](https://survey.stackoverflow.co/2023/#technology) was misinterpreted [here](https://github.com/AvaloniaUI/Avalonia/discussions/10387#discussioncomment-8186030). Visual Basic (.Net) and F# are preferred by 4.07% and 0.97% of the developers, respectively, so VB.NET is about 4 times more popular than F#.

@MikeCodesDotNET I understand your work is challenging due to very limited resources, but missing this kind of information can be harmful to Avalonia's growth. Thus, I think it would be a strategic mistake not to support VB.NET unless you have other statistics that motivate this decision. I hope you can learn from this and keep up the great work to take Avalonia further!

@DualBrain I agree that VB.NET should be supported. I know it may not be easy to understand, but Avalonia is growing very fast and it's not easy for Mike (nor would it be for anyone else) to handle it. Thus, we should try to be a bit more understanding. Finally, this is just my opinion, but please try to keep your answers short because they are a bit overwhelming to read and probably contributed to the locking of the original discussion.

**Comment by thevortexcloud** [comment]
I was also in the middle of writing a response when it was locked. So I am just going to put an abridged modified version of it here:

> I gave a pragmatic answer. I am also not here to have an argument about which language is best. In general, I personally do not have a problem with using VB. You are more than welcome to keep using it, and as stated by Mike, you are welcome to use it with Avalonia. You clearly care a lot about VB, which is a good thing. Having people be advocates for this sort of thing is how things can change.
>
> In any case, I am not a member of the Avalonia team. What gets officially supported is not up to me.
>
> Also as a Linux user I am used to being told "there is not enough demand", so I get where you are coming from.

As for your comparison compared to F#, I anecdotally have always felt that VB was actually bigger than F#. Most .NET devs in my experience are not even aware F# exists. But they have likely heard of some form of VB before.

**Comment by kekekeks** [comment]
VB.NET has the same level of support on the compiler level as C# and F#.

As of templates and code generators, somebody needs to create *maintain* those, which is currently out of team's limited capacity and there is no paid support customer who needs those.

We'd be happy to merge community-provided templates the moment we can be sure that those would continue to be supported by community (historically we've had more trouble with having to support stuff for ages than with initially implementing/merging it).

So, the way to get VB support is to follow the F# route of having community-backed repos for VB-related stuff first.

**Comment by CoolCoderSuper** [comment]
Avalonia is very usable with VB already. Here is a sample game with AvaloniaUI even includes WASM app with zero c# code https://github.com/CoolCoderSuper/SpaceInvaders. Does someone know where the c# source generator is for reference?

**Reply by maxkatz6** [reply]
> Does someone know where the c# source generator is for reference

https://github.com/AvaloniaUI/Avalonia/tree/master/src/tools/Avalonia.Generators

**Comment by DualBrain** [comment]
To be clear, I have no problem in when/if/how Avalonia UI supports VB. I feel that the answer of pointing to templates elsewhere is a perfectly acceptable answer and if it would have been left at that... I don't think there would be any issues to discuss.

My problem is that it seems totally acceptable to simply tell people that their choice is *wrong* and that they should abandon that choice because they think so and then try to validate such a response with questionable (at best) evidence.

What is clear here is that Avalonia UI has no knowledge of the VB community and zero interest in gaining any insight as such. And that's perfectly OK. What is not OK though is what amounts to insulting behavior toward what is literally a first-class language within the .NET ecosystem. That is not acceptable and, worse still, team members don't seem to understand that they might have cross a bit of a line - their actions, to them, are totally justifiable given the "evidence" at hand.

**Comment by DualBrain** [comment]
Feel free to take a look at some VB templates for Avalonia UI that are available (as source) over at https://github.com/CommunityVB/Templates/tree/main/src/avalonia

Will eventually make these available as a nuget package; but they are available now in source and certainly could use some testing/feedback.

**Reply by thevortexcloud** [reply]
Your link does not work, I am guessing it should be: https://github.com/CommunityVB/Templates/tree/main/src/avalonia

**Reply by DualBrain** [reply]
@thevortexcloud Thanks for pointing that out; adjusted the link.

**Comment by ivoryguard** [comment]
I have tested VB template for AvaloniaUI. It has some issue (duplicated .sln file) but it worked.
However, the lack of a source generator makes code behind functionality of xaml unavailable - it is not possible to define the control name with x:Name in xaml and refer it at .xaml.vb.

If source generator function for AvaloniaUI VB, AvaloniaUI will make it possible to bring the vb experience of WPF to cross platform development with AvaloniaUI.


One more thing:
I think that most of VB developers may either high level or beginners. Mid-level developers may have moved to C#.
High level VBer may be larger portion at now a days because the decrease of beginners.

VB is still heavily used in LOB applications. 
There may be less of a need for the latest .NET technolgies in LOB environments because they are often using legacy technologies (i.e. the .NET Framework instead of .NET 8) and prefer to stick with proven technologies. 

For this reason, it might be a mistake to assume that there are fewer VB users based on Stackoverflow. According to TIBO's language rankings, VB still ranks high, despite Microsoft's discrimination against VBers. (It is a political issue, not a technical issue)

As time goes on, even line-of-business applications require cross-platform development.

I created a simple AvaloniaUI sample project in C# and VB, and the Shared View/ViewModel project is a .NET Class library type;
There is no difference between VB and C# at .NET Class library type.

AvaloniaUI may be a good choice to develop cross platform apps that display data on the screen (which is what LOB applications are) rather than apps that use native resources. (Especially mobile first apps - the lack of navigation framework and so on)

If VB is better supported in Avalonia, I think a lot of potential old VBers would support AvaloniaUI project, and I'd like to help out a bit.

Best regards.

**Reply by CoolCoderSuper** [reply]
You can manually reference controls by their name with the FindControl method. But I agree source generators would be cool.

**Reply by ivoryguard** [reply]
Thank you for a good information.
I will test FindControl method. 

Using xaml element directly at code behind in the View level sometimes makes problems easy though MVVM pattern. 

The FindControl method can potentially serve as an alternative to source code generator, since the usage of FindControl in the MVVM pattern may be not frequent in general. 

What other limitations might there be in using VB with Avalonia?

**Reply by CoolCoderSuper** [reply]
If you are using the mobile backends there might be some tooling issues. In which case I would just have the main library in VB and then have the Android, IoS, and WASM projects just in C#.

**Reply by thevortexcloud** [reply]
If you are going to mix projects like that then there is nothing stopping you from putting all your views in a C# project (which will have proper source generator support) and your view models in a VB project. If you follow MVVM properly you likely wont have much code in your views code behind anyway. You can't mix .NET languages in the same project, but you can mix assemblies written in different languages as they all get compiled to the same code.

**Reply by CoolCoderSuper** [reply]
The reason would be for a generally unified code base, this doesn't matter for individual backends since you make changes in there very rarely. Also because I don't follow MVVM properly ;)

**Reply by ivoryguard** [reply]
Thank you for the reply.

I already have developed my products by writing most of the code in VB at class library projects and minimal C# code in an frontend ASP.NET Core MVC project containing html views, and it worked very well. 

I've also applied this approach to the Xamarin.Form project and the Blazor wasm project with great success.

However, I miss being able to build everything from scratch in VB like I used to do in WPF and Silverlight.

Anyway, the reason I chose Avalonia instead of Uno is that, like Silverlight, it offers full screen rendering for high performance, which is a significant advantage (like Flutter). Additionally, Avalonia has better support for VB than other platforms at least.

I believe that structurally, VB support in Uno might not be straightforward. Therefore, the VB support in Avalonia could be an advantage compared to other .NET cross-platform frameworks. By the way, there's also a project called OpenSilver, which officially supports VB.

Best regards.

**Comment by benyuz** [comment]
It is the power of avalonia developers to support vb.net or not, but I would like to say that although Microsoft has voluntarily abandoned vb.net, and even threatened not to support VB.NET in favor of C# and F#, in fact there are still a large number of developers in the vb language after all these years. Just because you don't speak doesn't mean you don't. I think this was a mistake on the part of Microsoft. C# and F# didn't live up to Microsoft's expectations, and if there was more support for VB then I think the contribution would have been better than C#.
